messages/ruby/lib/cucumber/messages/message/serialization.rb

```diff
@@ -0,0 +1,72 @@
+require 'cucumber/messages/message/utils'
+require 'json'
+
+module Cucumber
+  module Messages
+    class Message
+      include Cucumber::Messages::Message::Utils
+
+      module Serialization
+
+        ##
+        # Returns a new Hash formed from the message attributes
+        # If +camelize:+ keyword parameter is set to true, then keys will be camelized
+        # If +reject_nil_values:+ keyword parameter is set to true, resulting hash won't include nil values
+        #
+        #   Cucumber::Messages::Duration.new(seconds: 1, nanos: 42).to_h                                 # => { seconds: 1, nanos: 42 }
+        #   Cucumber::Messages::PickleTag.new(name: 'foo', ast_node_id: 'abc-def').to_h(camelize: true)  # => { name: 'foo', astNodeId: 'abc-def' }
+        #   Cucumber::Messages::PickleTag.new(name: 'foo', ast_node_id: nil).to_h(reject_nil_values: true)  # => { name: 'foo' }
+        #
+        # It is recursive so embedded messages are also processed
+        #
+        #   location = Cucumber::Messages::Location.new(line: 2)
+        #   Cucumber::Messages::Comment.new(location: location, text: 'comment').to_h  # => { location: { line: 2, :column: nil }, text: "comment" }
+        #
+
+        def to_h(camelize: false, reject_nil_values: false)
```

Comment:
What concrete use cases do we have for the `camelize` and `reject_nil_values` arguments?

Reply:
Those are used for json serialization.
`camelize` for the json using camelCase keys
`reject_nil_values` to avoid serializing null values as requested here: https://github.com/cucumber/common/pull/1605#discussion_r653065795

Reply:
I understand that, but why do we ever need to set `camelize` to `false` or `reject_nil_values` to `false`?

Was this flexibility added just in case someone else needs it? AFAICT we don't need it ourselves.

Reply:
> I understand that, but why do we ever need to set `camelize` to `false` or `reject_nil_values` to `false`?
> 
> Was this flexibility added just in case someone else needs it? AFAICT we don't need it ourselves.

Yes.

Because having a ruby hash with camelcase keys is not the ruby standard, and reject_nil_values to avoid data loss per default. So per default, the method has an intuitive use case.

Reply:
This speculative flexibility comes at a complexity and performance cost.

I assume you went with a reflection based serialization implementation in order to support this use case. If we implemented it without these option (always `camelize` and `reject_nil_values`), we wouldn't need the additional logic, and a static code generation would be sufficient (less complex and faster).

I think this is feature creep, but I'll rest my case here.

Reply:
> This speculative flexibility comes at a complexity and performance cost.

The whole deserialization thing, which is the big piece in the current PR, is based on speculation. At first we wanted to implement only the serialization, which is required for the DTOs to work properly, then you requested the deserialization just in case someone else would need it, because it would be a more comprehensive developer experience. We actually do not need it ourselves.

The addition of a `to_h` method to messages is at first for our own usage. Make it intuitive and public for third-parties developer is a bonus. Actually those options do not add any complexity to the code as they result in only 2 extra `if` statement.

```ruby
def to_h(camelize: false, reject_nil_values: false)
  resulting_hash = self.instance_variables.to_h do |variable_name|
    h_key = variable_name[1..-1]
    h_key = Cucumber::Messages::Message.camelize(h_key) if camelize

    h_value = prepare_value(
      self.instance_variable_get(variable_name),
      camelize: camelize,
      reject_nil_values: reject_nil_values
    )

    [ h_key.to_sym, h_value ]
  end

  resulting_hash.reject! { |_, value| value.nil? } if reject_nil_values
  resulting_hash
end

def to_h
  resulting_hash = self.instance_variables.to_h do |variable_name|
    h_key = variable_name[1..-1]
    h_key = Cucumber::Messages::Message.camelize(h_key)

    h_value = prepare_value(
      self.instance_variable_get(variable_name)
    )

    [ h_key.to_sym, h_value ]
  end

  resulting_hash.reject! { |_, value| value.nil? }
  resulting_hash
end
```

> I assume you went with a reflection based serialization implementation in order to support this use case. If we implemented it without these option (always camelize and reject_nil_values), we wouldn't need the additional logic, and a static code generation would be sufficient (less complex and faster).

There is no relation between using reflection, and the addition of those two options. I would have proposed the same options with generated static code.

Reply:
> The whole deserialization thing, which is the big piece in the current PR, is based on speculation. At first we wanted to implement only the serialization, which is required for the DTOs to work properly, then you requested the deserialization just in case someone else would need it, because it would be a more comprehensive developer experience. We actually do not need it ourselves.

Fair point. I think we should publish a library that will enable people to write consumers in Ruby. See https://github.com/cucumber/common/issues/1623 where someone was asking how to consume messages in Java.

It's speculative, but it also encouragement.

Reply:
Thank you @aslakhellesoy 🙂